src/main/java/teammates/common/util/ActivityLogEntry.java

```diff
@@ -1,631 +1,220 @@
-        return time;
+    public String getActionName() {
+        return actionName;
     }
     
-    public String getServletName() {
-        return servletName;
+    public String getUserRole() {
+        return userRole;
     }
     
-    public String getAction() {
-        return action;
+    public boolean isMasqueradeUserRole() {
+        return isMasqueradeUserRole;
     }
     
-    public String getRole() {
-        return role;
+    public String getUserName() {
+        return userName;
     }
     
-    public String getName() {
-        return name;
+    public String getUserGoogleId() {
+        return userGoogleId;
     }
     
-    public String getGoogleId() {
-        return googleId;
+    public String getUserEmail() {
+        return userEmail;
     }
     
-    public String getEmail() {
-        return email;
+    public boolean isTestingData() {
+        return userEmail.endsWith(Const.ActivityLog.TESTING_DATA_EMAIL_POSTFIX);
     }
     
-    public Long getTimeTaken() {
+    public static class Builder {
+        // Required parameters
+        private String actionServletName;
+        private String actionUrl;
+        private long logTime;
         
-        return timeTaken;
-    }
-    
-    /**
-     * Generates the ID for the log. If the googleId is unknown or unregistered,
-     * the email and course of the {@code student} will be used to construct the id.
-     * @param googleId the google ID
-     * @param student StudentAttributes object
-     * @return log ID
-     */
-    public String generateLogId(String googleId, StudentAttributes student, long time) {
-        return student == null ? generateLogId(googleId, null, null, time)
-                               : generateLogId(googleId, student.email, student.course, time);
-    }
-    
-    /**
-     * Generates the ID for the log. If the googleId is unknown or unregistered,
-     * the {@code email} and {@code course} will be used to construct the id.
-     * @param googleId the google ID
-     * @param email the email
-     * @param course the course
-     * @return log ID
-     */
-    public String generateLogId(String googleId, String email, String course, long time) {
-        boolean isUnregisteredStudent = (googleId.contentEquals("Unknown") || googleId.contentEquals("Unregistered"))
-                                        && email != null && course != null;
+        // Optional parameters - initialized to default values
+        private String actionName = Const.ActivityLog.UNKNOWN;
+        private long actionTimeTaken;
+        private String userRole = Const.ActivityLog.UNKNOWN;
+        private String userName = Const.ActivityLog.UNKNOWN;
+        private String userGoogleId = Const.ActivityLog.UNKNOWN;
+        private String userEmail = Const.ActivityLog.UNKNOWN;
+        private String logMessage = Const.ActivityLog.UNKNOWN;
+        private String logId = Const.ActivityLog.UNKNOWN;
+        private boolean isMasqueradeUserRole;
```

Comment:
isMasqueradeUserRole is not given a default value and withMasqueradeUserRole might not be called. This can create a null pointer and causes exception in ActivityLogEntry.generateLogMessage()